### ISSUE TITLE
Add AMP dtype

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#424]
This depends on #423.

This would let each operation/optimizer choose which precision it uses.

The exact interface needs to be thought out, but here's a potential sketch:

```rust
struct AMP<Lo: Dtype, Hi: Dtype> { ... }

let q: Tensor<Rank1<5>, AMP<f16, f32>, _> = dev.zeros();
```

Things to figure out:
- How would certain layers choose an actual dtype from the two options in Amp? trait implementations?
- Is this even possible? Right now Dtype is just used directly for storage (ie. `Vec<Dtype>`), but this proposal is kind of a Dtype container. Does `Vec<AMP<f16, f32>>` even make sense?
- What are the generics on AMP? Are they even needed (i.e. implied f16/f32)?
- Need to write down exactly what operations are f16/f32, and what tensor/gradients are defaulted to (I think the lower precision).

**Comment by opfromthestart** [comment]
Would an enum work for this? I imagine it could be something like
```rust
#[derive(Clone, Copy)]
enum AMP<Lo: Dtype, Hi: Dtype> {
    Low(Lo),
    High(Hi),
}

impl<Lo: Dtype, Hi: Dtype> Amp<Lo, Hi> {
    fn to_low(self) -> Lo;
    fn to_high(self) -> Hi;
    fn from<E: Dtype>(val: E) -> Self;
}
```

**Comment by coreylowman** [comment]
I think probably not actually if I'm understanding AMP properly. I think we'd need:
```rust
#[repr(transparent)]
struct AMP(half::f16);
```

Which would mean that on Cuda side, we are just storing halves. Then any cuda kernels could accept inputs of `const *half` directly, and if they wanted to do mixed precision each kernel can select that (since we'd have sum_fwd_f16 and sum_fwd_amp as separate kernels).

It's not clear to me what would be stored Cuda side if we did enums

**Comment by jafioti** [comment]
@coreylowman What would need to be done to push this across the line? I'd like to help out, whatever it takes to get this in cause I'd like to train in fp16!

**Comment by coreylowman** [comment]
@jafioti not much work has been done on this so far. I think first step is getting the type set up so you can impl Dtype for it. Then being able to instantiate tensors with AMP dtype, and then try writing a kernel for it and making sure tests can compile.

**Comment by coreylowman** [comment]
Okay so the main parts of this are done. I think the main part of this left is the gradient scaling part. It looks like that involves:
1. Scaling the loss by some value (which is pretty easy)
2. Unscaling the gradients before they are used (not as easy)

I'm a bit unclear how to do the unscaling in an easy way. Does anyone have experience with using gradient scaling? If it's pretty much always scaled by some standard value, then maybe we can just hardcode this for now.

**Comment by coreylowman** [comment]
I'm tempted to leave gradient scaling for a separate issue, then we can close this one out and discuss more specifics there
